Implement performance interface extension for PerformanceResourceTiming

aditj opened the pull request:
<!-- Please describe your changes on the following line: -->
UPDATE: Built it on my local machine and removed the errors occurring during compilation.

Implemented the following : 
 a) Added the maxSize Field to the PerformanceEntryList structure.
 b) Implemented the clearResourceTimings, setResourceTimingBufferSize, onresourcetimingbufferfull fuctions for the Performance Structure according to the Interface Specification given at : https://w3c.github.io/resource-timing/#sec-extensions-performance-interface

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix #22307
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22431)
<!-- Reviewable:end -->

aditj committed:
Implement performance interface extension

Implemented the following :
 a) Added the maxSize Field to the PerformanceEntryList structure.
 b) Implemented the clearResourceTimings, setResourceTimingBufferSize, onresourcetimingbufferfull fuctions for the Performance Structure according to the Interface Specification given at : https://w3c.github.io/resource-timing/#sec-extensions-performance-interface

aditj commented:
The local version is building fine but the TravisCI build is failing because the functions implemented are never being used , but that should be more of a warning and not an error. :/

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -113,6 +114,25 @@ impl PerformanceEntryList {
     }
+    //clears the buffer
+    fn clearResourceTimings(&mut self){
+         self.entries.clear();
```
This would clear all the registered entries. We don't want that. What we want is to clear only the ones with `PerformanceResourceTiming` type.

We have the `clear_entries_by_name_and_type` to help with this task. So you could do something like:

```rust
self.clear_entries_by_name_and_type(None, Some(DOMString::from("resource")));
```

This method belongs to the `Performance` interface, and so it should be implemented in the `Performance` struct.

ferjm reviewed (changes requested):
Thanks for the PR.

There are existing [Web Platform Tests](https://web-platform-tests.org/introduction.html) that you can use to test your patches:

- [tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_eventually.html](https://github.com/servo/servo/blob/42ca43e51fd705d6c0c915c699abf65b4b15de78/tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_eventually.html)
- [tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_when_shrink_buffer_size.html](https://github.com/servo/servo/blob/16bdf9225dd979feeea1580f78041016ce41643f/tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_when_shrink_buffer_size.html)
- [tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_when_populate_entries.html](https://github.com/servo/servo/blob/d657a6c6e885cc3bc58f82e8616b50d358fe8ece/tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_when_populate_entries.html)
- [tests/wpt/web-platform-tests/resource-timing/resource_timing_store_and_clear_during_callback.html](https://github.com/servo/servo/blob/d657a6c6e885cc3bc58f82e8616b50d358fe8ece/tests/wpt/web-platform-tests/resource-timing/resource_timing_store_and_clear_during_callback.html)
- [tests/wpt/web-platform-tests/resource-timing/clear_resource_timing_functionality.html](https://github.com/servo/servo/blob/16bdf9225dd979feeea1580f78041016ce41643f/tests/wpt/web-platform-tests/resource-timing/clear_resource_timing_functionality.html)

We have [documentation](https://github.com/servo/servo/tree/master/tests/wpt#running-the-tests) about how to run these tests locally.

I would also recomment to run `./mach fmt` before submitting updated patches. This command uses [rustfmt](https://github.com/rust-lang/rustfmt) to format your code. The CI process checks this formatting and [rejects](https://travis-ci.org/servo/servo/jobs/466887188#L452) any new code with unexpected formatting.

The code needs several changes:

- You are trying to implement an extension to the `Performance` interface. DOM interfaces are defined via [Web IDL](https://en.wikipedia.org/wiki/Web_IDL). In particular, you have to modify the [Performance.webidl](https://github.com/servo/servo/blob/42ca43e51fd705d6c0c915c699abf65b4b15de78/components/script/dom/webidls/Performance.webidl) file to add the new methods and the event handler defined by the spec.

```webidl
partial interface Performance {
  void clearResourceTimings ();
  void setResourceTimingBufferSize (unsigned long maxSize);
  attribute EventHandler onresourcetimingbufferfull;
};
```

- The previous changes to the `Performance.webidl` file will add new methods to the `PerformanceBindingMethods` trait that you will need to implement [here](https://github.com/servo/servo/blob/42ca43e51fd705d6c0c915c699abf65b4b15de78/components/script/dom/performance.rs#L304). So you'll need to add something like:

```rust
fn ClearResourceTimings(&self) {
   // implementation goes here.
}

fn SetResourceTimingBufferSize(&self, max_size: u32) {
  // implementation goes here.
}

event_handler!(resourcetimingbufferfull, GetOnresourcetimingbufferfull, SetOnresourcetimingbufferfull);
```

- I added some more comments to the code inline.

aditj commented on components/script/dom/performance.rs:
```diff
@@ -56,11 +56,12 @@ const INVALID_ENTRY_NAMES: &'static [&'static str] = &[
 #[derive(JSTraceable, MallocSizeOf)]
 pub struct PerformanceEntryList {
     entries: DOMPerformanceEntryList,
+    maxSize:usize,
```
@ferjm Thanks for the help!
Just a query
a) cant we implement resource_timing_secondary_entries as DomRefCell<PerformanceEntryList> instead of DomRefCell<Vec<DomRoot<PerformanceEntry>>>?
b) Also If we  are implementing it as DomRefCell<Vec<DomRoot<PerformanceEntry>>> , can you guide me on how to initialise it in the new_inherited function? :sweat_smile:  Thanks again :smile:

ferjm replied:
> a) cant we implement resource_timing_secondary_entries as
DomRefCell<PerformanceEntryList> instead of DomRefCell<Vec<DomRoot<PerformanceEntry>>>?

I don't think we need the full power of a PerfomanceEntryList in this case, a Vec should be sufficient for our needs.

> b) Also If we are implementing it as DomRefCell<Vec<DomRoot>> , can you guide me on how to initialise it in the new_inherited function?

```rust
resource_timing_secondary_entries: DomRefCell::new(Vec::new()),
```

aditj commented:
Also how do I write the interface in webdl such that the PerformanceBindings that is created has a mutable parameter of self in setResourceTimingBufferSize ?

ferjm commented:
> Also how do I write the interface in webdl such that the PerformanceBindings that is created has a mutable parameter of self in setResourceTimingBufferSize ?

AFAIK there's no way to do that. You'll need to use [interior mutability](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html) in this case.

```rust
resource_timing_buffer_size: Cell<usize>,
```

aditj committed:
Executed the required changes

Update servo-media. Fixes discarded frames during player setup

Update web-platform-tests to revision 6087baf4a83e7953112242be9fd6e719797ebdf6

Bug 1509027 - Use more restricted scope for Treeherder routes

https://bugzilla.mozilla.org/show_bug.cgi?id=1509027

Cleanup constellation after nll

Use the base Ubuntu distro instead of LinuxMint in './mach bootstrap'

Fixup the issue related with an incorrect distro version string handling

Kill test processes whenever any error occurs

Update web-platform-tests to revision 5349c6a6327372f856ecbe07d863d468a4236726

Minor pull request template cleanup

Introduce NavigationPreloadManager for ServiceWorkerRegistration

Update web-platform-tests to revision 2d5385793ade85b7268ee86db4535a1baeb508ab

Cancel animations that affect nodes that do not participate in layout.

Add a test for dealing with multiple image loading tasks when the list of available images is usable.

Add info about building on Linux Mint to README

Remove redundant `.clone()`s

Update web-platform-tests to revision 912d5081b62d6e6a2f847935c82722e31cca7a1f

Add debug output for webdriver commands.

Delay setting the constellation's active browser id until the context exists.

Add headless support to servodriver harness.

Add automated test for servodriver harness.

Make SSL tests work in servodriver.

performance.rs

Added the specification link

Changed the queue_entry function

CYBAI commented:
On Travis

```
$ ./mach test-tidy --no-progress --all
 0:11.90 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/mozilla/meta/MANIFEST.json
 0:12.04 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/metadata/MANIFEST.json
 0:16.25 INFO Diffing old and new manifests /home/travis/build/servo/servo/tests/wpt/webgl/meta/MANIFEST.json
Checking the config file...
Checking the wpt manifest file...
Checking directories for correct file extensions...
Running the WPT lint on ./tests/wpt/web-platform-tests/...
Running the WPT lint on ./tests/wpt/mozilla/tests/...
Checking files for tidiness...
./components/script/dom/performance.rs:416: method declared in webidl is missing a comment with a specification link
./components/script/dom/performance.rs:422: method declared in webidl is missing a comment with a specification link
./components/script/dom/performance.rs:426: method declared in webidl is missing a comment with a specification link
./components/script/dom/webidls/Performance.webidl:26: trailing whitespace
Running the dependency licensing lint...
info: downloading component 'rustfmt-preview'
info: installing component 'rustfmt-preview'
The command "./mach test-tidy --no-progress --all" exited with 1.
```

ferjm reviewed (changes requested):
Thanks for the updated patch. This is a good improvement.

There are still some missing steps:

- We need to implement the steps described in the [_add a PerformanceResourceTiming entry_](https://w3c.github.io/resource-timing/#sec-extensions-performance-interface) paragraph of the spec. We are currently adding all types of performance entries [here](https://github.com/servo/servo/blob/master/components/script/dom/performance.rs#L209), including the resource timing ones. So that's the method that we need to modify. Try following the steps described in the spec and don't hesitate to ask if you don't know how to implement any of them.

- I recommend implementing independent methods for the other three algorithms described in the spec, so we can use them while implementing the previous step:

```rust
fn can_add_resource_timing_entry(&self) -> bool {...}
fn copy_secondary_resource_timing_buffer(&self) {...}
fn fire_buffer_full_event(&self) {...}
```

aditj commented on components/script/dom/webidls/Performance.webidl:
```diff
@@ -23,6 +23,7 @@ partial interface Performance {
   PerformanceEntryList getEntriesByType(DOMString type);
   PerformanceEntryList getEntriesByName(DOMString name,
                                         optional DOMString type);
+  
```
Which line?

ferjm replied:
26

aditj commented:
Where do I implement the secondary methods :
```
fn can_add_resource_timing_entry(&self) -> bool {...}
fn copy_secondary_resource_timing_buffer(&self) {...}
fn fire_buffer_full_event(&self) {...}
```
?
Also where do I put a check for can_add_resource_timing_entry?

ferjm commented:
They should all be part of the `Performance` struct implementation.

aditj commented:
> They should all be part of the `Performance` struct implementation.

Do I need to make these functions public or is it okay for them to be private since these will be accessed inside from Performance's member functions only?

ferjm commented:
Private is fine.

aditj commented:
And is popping from secondary buffer fine? 
( The specs mention:

aditj commented:
Also in the specs :
`Decrement resource timing secondary buffer current size by 1`
So should the Performance structure also have  secondary buffer size as a member?
or is caching it not important?

aditj commented:
Also how do we fire events at objects ? Is there some documentation that can help?

jdm commented:
Look for uses of [Event.fire](https://doc.servo.org/script/dom/event/struct.Event.html#method.fire).

ferjm commented:
> And is popping from secondary buffer fine?
>( The specs mention:
>Let entry be the oldest PerformanceResourceTiming in resource timing secondary buffer.
>Add entry to the end of performance entry buffer.)

That means that you need to process the elements in the same order that you added them to the buffer (FIFO, so it's a queue). And actually, there's a better data structure for this: [VecDeque](https://doc.rust-lang.org/std/collections/struct.VecDeque.html).

> Also in the specs :
> Decrement resource timing secondary buffer current size by 1
> So should the Performance structure also have secondary buffer size as a member?
> or is caching it not important?

That would be the length of the `resource_timing_secondary_entries` collection that you added [here](https://github.com/servo/servo/pull/22431/files#diff-f0f83a3426dbfdaa9799c211fe72847bR143). I don't think we need to cache that value in this case. We need to cache the number of resource timing entries in the main [entries](https://github.com/servo/servo/pull/22431/files#diff-f0f83a3426dbfdaa9799c211fe72847bR136) buffer, because that buffer contains all types of performance entries (not only resource timing entries), and it is easier to cache the number of entries of type `resource` than inspecting the entire buffer every time we want to know that number. But in this case, we should be fine just checking `resource_timing_secondary_entries.len()`.

aditj commented:
About the Event Firing again, Do I create an event object ? Or do i use the .fire method of a pre existing method?

jdm commented:
For simple events like the ones in the specification, you can use the [EventTarget::fire_event](https://doc.servo.org/script/dom/eventtarget/struct.EventTarget.html#method.fire_event) method. Look for other uses in component/script as models.

aditj commented:
> 
> > Also in the specs :
> > Decrement resource timing secondary buffer current size by 1
> > So should the Performance structure also have secondary buffer size as a member?
> > or is caching it not important?
> 
> That would be the length of the `resource_timing_secondary_entries` collection that you added [here](https://github.com/servo/servo/pull/22431/files#diff-f0f83a3426dbfdaa9799c211fe72847bR143). I don't think we need to cache that value in this case. We need to cache the number of resource timing entries in the main [entries](https://github.com/servo/servo/pull/22431/files#diff-f0f83a3426dbfdaa9799c211fe72847bR136) buffer, because that buffer contains all types of performance entries (not only resource timing entries), and it is easier to cache the number of entries of type `resource` than inspecting the entire buffer every time we want to know that number. But in this case, we should be fine just checking `resource_timing_secondary_entries.len()`.
 
I think we need to define a  secondary buffer size  member since the `resource_timing_secondary_entries structure does not have a length attribute?
Also is there anyway to set the flags to false other than providing a mutable reference to the function?

jdm commented:
Use Cell or DomRefCell to allow murating members without a mutable self reference.

aditj commented:
How do I deref a root object ?
I am still stuck on the event thing, what does an eventhandler macro do? Where do I create the event resourcetimingbufferfull? 
Is there any other file where a event has been created and fired at the target? (for reference purpose)
Thanks for the help :).

jdm commented:
You can find the macro definition in components/script/dom/macros.rs. If you search for code that yses Event::new or fire you should find existing code that can help you understand.

jdm commented:
And you dereference a root object with the . operator to access the inner type. It's like Box or Rc or other smart pointer types.

ferjm commented:
> I think we need to define a secondary buffer size member since the `resource_timing_secondary_entries structure does not have a length attribute?

I am fairly sure that `resource_timing_secondary_entries` has a `len()` method. It is a `DomRefCell<VecDequeue<...>>` and as a [VecDequeue](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) it has a [len](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.len) method. You need to [borrow](https://doc.servo.org/script/dom/bindings/cell/struct.DomRefCell.html#method.borrow) the collection from the `DomRefCell` to gain access to its methods.

aditj commented:
> And you dereference a root object with the . operator to access the inner type. It's like Box or Rc or other smart pointer types.
The dot operator is not working. 
It gives the following error.

Also a side question : Can I commit and push even if the code has some errors or is it discouraged?

aditj commented:
> And you dereference a root object with the . operator to access the inner type. It's like Box or Rc or other smart pointer types.

 I tried the dot operator with deref function but it is giving the following error:
```
 pub struct Root<T: StableTraceObject> {                                   
    | ------------------------------------- method `deref` not found for this 
```
 
Just another small question : Can I commit and push the code even if it has errors or is it discouraged ?

aditj commented:
```conflicting implementations of trait `dom::performance::_IMPL_DOMOBJECT_FOR_Performance::ShouldNotImplDomObject` for type `((), dom::event::Event)```
What does this error mean ?

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +298,31 @@ impl Performance {
+	}
+	fn fire_buffer_full_event(&self) {
+		while self.resource_timing_secondary_entries.borrow_mut().len() > 0 {
+			if self.can_add_resource_timing_entry()==false{
```
No need to explicitly compare to `false`. You can do
```rust
if !self.can_add_resource_timing_entry() {
```

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +298,31 @@ impl Performance {
+		}
+	}
+	fn fire_buffer_full_event(&self) {
+		while self.resource_timing_secondary_entries.borrow_mut().len() > 0 {
```
No need to mutably borrow here, and you can use the `is_empty` method.

```rust
while !self.resource_timing_secondary_entries.borrow().is_empty() {
```

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +298,31 @@ impl Performance {
+		self.queue_entry(&(*(self.resource_timing_secondary_entries.borrow_mut().pop_front().unwrap())),true);
+		}
+	}
+	fn fire_buffer_full_event(&self) {
```
We are missing a few steps from the spec here (i.e. steps 1.1, 1.4, 1.5). Please, try to follow the spec as much as possible.

aditj commented:
The local build does not give a specific error.
Did the ./mach fmt
Why are the two checks failing? :confused:

ferjm commented:
That's `./mach test-tidy` failing.

The [errors](https://travis-ci.org/servo/servo/jobs/469619300#L464) are:

```
./components/script/dom/performance.rs:484: method declared in webidl is missing a comment with a specification link
./components/script/dom/performance.rs:492: method declared in webidl is missing a comment with a specification link
./components/script/dom/performance.rs:498: method declared in webidl is missing a comment with a specification link
```

ferjm commented:
The linter is complaining about the missing spec links in the implementation (`components/script/dom/performance.rs`), not in the interface definition (`components/script/dom/webidls/Performance.webidl`).

Try not to use CI as testing playground, please. You can test locally by running `./mach test-tidy`.

aditj commented:
./mach test-tidy is running fine locally
./mach build -d is building without any errors
The code has been formatted using ./mach fmt 
The travisCI is passing without any errors.
 

- Does the code require any more changes?
- Is the commenting appropriate?
- Please tell if any other logical or less obvious errors are there? :)

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+            .is_empty()
+        {
+            let no_of_excess_entries_before =
+                self.resource_timing_secondary_entries.borrow_mut().len();
```
No need to mutably borrow here.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+            let no_of_excess_entries_before =
+                self.resource_timing_secondary_entries.borrow_mut().len();
+            let mut no_of_excess_entries_after =
+                self.resource_timing_secondary_entries.borrow_mut().len();
```
There is no need to get the length again until __after__ step 3.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+            );
+        }
+    }
+    fn fire_buffer_full_event(&self) {
```
nit: one thing that helps reading the code and comparing it with the spec is adding the link to the specified algorithm above the function definition and a comment identifying the different steps. For example:
```rust
// `fire a buffer full event` paragraph of
// https://w3c.github.io/resource-timing/#sec-extensions-performance-interface
fn fire_buffer_full_event(&self) {
    // Step 1.
            while !self
            .resource_timing_secondary_entries
            .borrow()
            .is_empty()
    {
        // Step 1.1.
        let no_of_excess_entries_before =
                self.resource_timing_secondary_entries.borrow().len();
    
[...]
```

aditj replied:
Cool! Should I document all the methods I wrote?

ferjm replied:
Yes, please.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+        while !self
+            .resource_timing_secondary_entries
+            .borrow_mut()
+            .is_empty()
```
Same as before, there is no change to the collection here, so we don't need to mutably borrow it.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+                self.resource_timing_secondary_entries.borrow_mut().len();
+            if !self.can_add_resource_timing_entry() {
+                self.eventtarget
+                    .fire_event(atom!("resourcetimingbufferfull"));
```
```rust
self.upcast::<EventTarget>()
      .fire_event(atom!("resourcetimingbufferfull"));
```

jdm replied:
https://doc.servo.org/script/dom/index.html#inheritance and https://doc.servo.org/script/dom/bindings/inheritance/trait.Castable.html should make it more clear.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+            if !self.can_add_resource_timing_entry() {
+                self.eventtarget
+                    .fire_event(atom!("resourcetimingbufferfull"));
+                self.copy_secondary_resource_timing_buffer();
```
This needs to be outside of the `if` block as the spec does not mention that copying to the secondary buffer or the rest of the steps depend on the ability to add timing entries.

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+                no_of_excess_entries_after =
+                    self.resource_timing_secondary_entries.borrow_mut().len();
+            }
+            if no_of_excess_entries_after >= no_of_excess_entries_before {
```
nit: let's write this the way that it is specified :)

```rust
if no_of_excess_entries_before <= no_of_excess_entries_after
```

ferjm commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+    fn should_queue_resource_entry(&self, entry: &PerformanceEntry) -> bool {
+        if self.can_add_resource_timing_entry() &&
+            !self.resource_timing_buffer_pending_full_event.get()
+        {
```
In this case we should only increase `resource_timing_buffer_current_size` and return true, so the resource entry is queued by the `queue_entry` function. Otherwise, we should implement the rest of the steps (3, 4 and 5 of the spec) here before returning `false`.

ferjm reviewed (changes requested):
Looking better!

jdm commented:
These changes should make `./mach test-wpt tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_eventually.html` actually run to completion.

aditj commented on components/script/dom/performance.rs:
```diff
@@ -223,19 +232,24 @@ impl Performance {
                 .borrow_mut()
-                .entries
-                .push(DomRoot::from_ref(entry));
+                .push_back(DomRoot::from_ref(entry));
```
> I am afraid that these changes break the insertion of entries that are not resource timing entries.

Should I add the code for   insertion of entries  that I removed and call the function should_queue_resource_entry if the entry is of type resource?

aditj commented:
> These changes should make `./mach test-wpt tests/wpt/web-platform-tests/resource-timing/resource_timing_buffer_full_eventually.html` actually run to completion.

Is there a way to build a single file locally , other wise the local builds take up a lot of time.

jdm commented:
Unfortunately no.

jdm commented:
`./mach check` can be used to identify any compilation errors without building everything, however.

aditj commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +287,70 @@ impl Performance {
+                    .unwrap())),
+                true,
+            );
+        }
```
I think we still need to unwrap as entry is an option and the compiler is giving the error
```
self.queue_entry(entry, true);                            
    |                                  ^^^^^ expected reference, found enum `std::option::Option`
```

KiChjang replied:
There's a typo in @ferjm's code: it should be `if let Some(ref entry) = entry {`.

KiChjang replied:
Also, I think he really meant this:

```rust
loop {
    if self.can_add_resource_timing_entry() {
        break;
    }
    let entry = self
        .resource_timing_secondary_entries
        .borrow_mut()
        .pop_front();
    if let Some(ref entry) = entry {
        self.queue_entry(entry, true);
        continue;
    } else {
        break;
    }
}
```

aditj commented:
Is there a reason AppVeyor Build fails? Cause even my local build is failing(and has failed earlier a few times) without giving a specific reason.
I have formatted the code and the test-tidy runs fine.

jdm commented:
That's #22476.

aditj commented:
Do I need to change some more files(except for https://github.com/servo/servo/blob/master/components/atoms/static_atoms.txt ) for the event atom than I inserted ? 

My code is not building locally and no specific errors are thrown except hundreds of lines of code saying undefined reference to JS objects.

aditj commented:
> I am not sure why you are getting these build errors. According to Travis the current patch builds fine. Could you update the PR with your latest changes, please?
My local branch is at par with the remote branch , also the errors on my local build are related to my operating system .
Is there any way I can locally run tests on my code or does TravisCI automatically does that?

ferjm commented:
TravisCI only runs unit tests, what you want to run are Web Platform Tests and you will need to be able to build locally to do that, unfortunately.

If you're having problems successfully building, please file a separate issue with the details of the build errors that you are observing, so we can try to determine the problem.

aditj commented:
> TravisCI only runs unit tests, what you want to run are Web Platform Tests and you will need to be able to build locally to do that, unfortunately.
> 
> If you're having problems successfully building, please file a separate issue with the details of the build errors that you are observing, so we can try to determine the problem.

 UPDATE: The local build is working fine but the wpt test is failing :sweat_smile: 
The gist file for the failed test :
https://gist.github.com/aditj/0ed3e3bbac560644e87c11910a65c2e4#file-gistfile1-txt

ferjm reviewed (changes requested):
Almost there!

aditj commented on components/script/dom/performance.rs:
```diff
@@ -273,10 +286,69 @@ impl Performance {
+        } else {
+            return false;
+        }
+    }
```
Thanks for help with the commenting! :)

aditj commented on components/script/dom/performance.rs:
```diff
@@ -223,19 +232,23 @@ impl Performance {
-                .entries
-                .push(DomRoot::from_ref(entry));
+            if entry.entry_type() == "resource" && self.should_queue_resource_entry(entry) {
+                return;
```
If I implement this then the order of steps would change (and would become not according to those mentioned in the steps).

aditj replied:
And btw where are the specs for queue entry function given ( I'll use them to document the code accordingly)

ferjm reviewed (changes requested):
Looks good! I think there are only a couple of things left to do:

- Update the tests expectations:
```
./mach test-wpt tests/wpt/web-platform-tests/resource-timing --log-raw /tmp/servo.log
./mach update-wpt /tmp/servo.log
```
- Squash all the commits

aditj commented:
I did a blunder I guess,I squashed all the commits, and then merged the master branch with the origin which changed some files , Can I remove the last commit somehow? :cry:

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1545929471009.

ferjm commented:
Hmm... I don't know exactly what happened here, but it seems that you did a bad merge. Is there any chance that you can push your original patches? I can guide you through the squash and rebase process.

aditj commented:
I force pushed the original patch. The tests I ran changed some files due to which some conflicting files have changed. 
(Also heads up on clearResourceTimings test is not passing .)

aditj commented:
How to I remove the merge on the remote repo?

aditj committed:
patch-1 restored

aditj commented:
Can you please guide me through the process ( if there is any) to revert the merge I did and to push the changes I have done locally ?

ferjm commented:
Looking at the current set of commits, it seems that most part of your changes are lost after force pushing.

I have a local clone of your branch with all the commits before [my last review comment](https://github.com/servo/servo/pull/22431#pullrequestreview-188082073). I can force push here after [squashing and rebasing](https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/), so you can run and update the expectations for the WPTs.

aditj committed:
Merge

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1546337469877.
